### DISCUSSION TITLE
ImportError while importing pyswmm

### DISCUSSION THREAD

**Discussion by argygt** [pyswmm/pyswmm#440] [Q&A] [answered]
Hello, 
I would like to read water quality data at nodes from SWMM. When I import pyswmm, I still face this error:
ImportError: DLL load failed while importing _solver: The specified procedure could not be found.
I have red in stackoverflow that this issue have been corrected with the latest versions. However I still face this issue.   
I am trying both pyswmm versions 1.4 and 1.2.
I use swmm-toolkit 0.14.0, SWMM 5.1.014,  Anaconda3 2020.11 (Python 3.8.5 65-bit)
and Jupiter notebook 6.1.4.
Suggestions about setting environment variable also did not work.
I am new in python, how can I resolve this issue?

**Comment by karosc** [comment]
@argygt , since you're getting a DLL import error, I assume you're running windows? Can you post the output of the following code?

```
import os
import sys
print(sys.prefix)
print(os.path.exists(os.path.join(sys.prefix, 'conda-meta')))

```

**Reply by argygt** [reply]
I am running Windows 10 Home 64bit. 
The output of the above code is: 
C:\Users\Argyro\anaconda3
True

**Comment by karosc** [comment]
can you try importing pyswmm in a fresh conda environment?

```
conda update conda -n base
conda create -n pyswmm_env python=3.8
conda activate pyswmm_env
pip install pyswmm==1.4
```

**Reply by argygt** [reply]
I create a new environment, but I still get the same error.
ImportError                               Traceback (most recent call last)
Cell In[1], line 1
----> 1 from pyswmm import Simulation, Nodes, Links, Output

File ~\anaconda3\envs\pyswmm1_4_2_env\lib\site-packages\pyswmm\__init__.py:11
      8 """Python Wrapper for Stormwater Management Model (SWMM5)."""
     10 # Local imports
---> 11 from pyswmm.links import Link, Links
     12 from pyswmm.lidcontrols import LidControls, LidControl
     13 from pyswmm.lidgroups import LidGroups, LidGroup, LidUnit

File ~\anaconda3\envs\pyswmm1_4_2_env\lib\site-packages\pyswmm\links.py:12
      9 from swmm.toolkit import shared_enum
     11 # Local imports
---> 12 from pyswmm.swmm5 import PYSWMMException
     13 from pyswmm.toolkitapi import LinkParams, LinkResults, LinkPollut, LinkType, ObjectType
     16 class Links(object):

File ~\anaconda3\envs\pyswmm1_4_2_env\lib\site-packages\pyswmm\swmm5.py:20
     17 from datetime import datetime
     19 # Third party imports
---> 20 from swmm.toolkit import solver
     22 # Local imports
     23 import pyswmm.toolkitapi as tka

File ~\anaconda3\envs\pyswmm1_4_2_env\lib\site-packages\swmm\toolkit\solver.py:18
     16 # Import the low-level C/C++ module
     17 if __package__ or "." in __name__:
---> 18     from . import _solver
     19 else:
     20     import _solver

ImportError: DLL load failed while importing _solver: The specified procedure could not be found.

I have to mention that I did not update the base environment because I am afraid that this may cause problems to other projects that I have already run in the base environment.

**Reply by karosc** [reply]
Could you paste the output of `conda info` here?

**Reply by karosc** [reply]
Also please post the output of the following commands in your fresh conda environment

```
 python -c "import sys; print(sys.version)"
 python -c "import platform; print(platform.system())"

```

**Reply by argygt** [reply]
active environment : pyswmm1_4_2_env
    active env location : C:\Users\Argyro\anaconda3\envs\pyswmm1_4_2_env
            shell level : 1
       user config file : C:\Users\Argyro\.condarc
 populated config files : C:\Users\Argyro\.condarc
          conda version : 4.10.1
    conda-build version : 3.20.5
         python version : 3.8.5.final.0
       virtual packages : __win=0=0
                          __archspec=1=x86_64
       base environment : C:\Users\Argyro\anaconda3  (writable)
      conda av data dir : C:\Users\Argyro\anaconda3\etc\conda
  conda av metadata url : https://repo.anaconda.com/pkgs/main
           channel URLs : https://repo.anaconda.com/pkgs/main/win-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/win-64
                          https://repo.anaconda.com/pkgs/r/noarch
                          https://repo.anaconda.com/pkgs/msys2/win-64
                          https://repo.anaconda.com/pkgs/msys2/noarch
          package cache : C:\Users\Argyro\anaconda3\pkgs
                          C:\Users\Argyro\.conda\pkgs
                          C:\Users\Argyro\AppData\Local\conda\conda\pkgs
       envs directories : C:\Users\Argyro\anaconda3\envs
                          C:\Users\Argyro\.conda\envs
                          C:\Users\Argyro\AppData\Local\conda\conda\envs
               platform : win-64
             user-agent : conda/4.10.1 requests/2.24.0 CPython/3.8.5 Windows/10 Windows/10.0.19041
          administrator : False
             netrc file : None
           offline mode : False


Note: you may need to restart the kernel to use updated packages.

--------

3.8.16 (default, Mar  2 2023, 03:18:16) [MSC v.1916 64 bit (AMD64)]

-------
Windows

**Reply by karosc** [reply, marked as answer]
Hmmm, your version of conda is several years old and I'm wondering if you should update it. FWIW, running `conda update conda -n base` should only update the conda package manager and not your python libraries in the base environment. If you are really concerned about it, you can try to [clone your base environment](https://conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#cloning-an-environment) into a proper conda environment prior to updating conda. I would try the following:

```
conda create --name cloned_base --clone base
conda update conda -n base
conda create --name pyswmm python -c conda-forge
conda activate pywmm
pip install pyswmm
```

FWIW, I run Windows 10 64-bit and strictly use conda as my preferred python distribution (although I use miniforge rather than the more bloated anaconda) and do not run into any issues like yours.

**Reply by argygt** [reply]
Updating conda finally works for me. And older projects also run properly at the cloned base environment. 
Thank you for your help!!
